Feat/ Tabs

Nullably opened the pull request:
**Before submitting this PR, please make sure**:

* [x] Your code builds clean without any errors or warnings
* [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
- Create new tab feature
- Fix tests failed on `test_interface_utils.py`, this is not part of the feature but for some reasons (unknowns) the tests failed after changes on this PR is introduced.
- Tab navigation by keyboard is tricky as many keys are reserved to OS and the application which Hummingbot is running under. I have set this to `escape` + `left` (or `right`). On Mac OS, this is `command` + left, right, if this doesn't work for you (depending on if you are using Terminal, iTerm or any others), you can try `esc` + left, right (you might need to press them at the same time). For Windows and Unix, this translate to  Alt + left, right.

**Tests performed by the developer**:
- Test tabs on MacOS

**Tips for QA testing**:
- Control + N for next tab and Control + P for previous tab.
- There is a tab_example command to show how a quick example if someone wants to create a new tab, it is also for us to test tabs creation/navigation when multiple tabs are created
- You can try to change the colour of the Hummingbot log_pane and terminal_primary to see how tabs colours change
- You can try create a new tab by simply copying tab_example.py file, change the file, class and command_name.


┆Issue is synchronized with this [Clickup task](https://app.clickup.com/t/1tre44m) by [Unito](https://www.unito.io)

Nullably commented:
@luisfilipeoliveira 
Can you please take a look into this PR? Are the layout and flow ok? I modified some features from what you suggested.

![image](https://user-images.githubusercontent.com/37262506/139658363-05e134db-1e64-4427-aad8-62b5644f5558.png)

luisfilipeoliveira commented:
@Nullably I have the following points for comment regarding this PR:

1. The paddings in the log-pane are missing, but I believe that a dependency on this PR: #4579 is missing. Am I correct? As @aarmoa mentioned in today's call (November 1).

2. The tab line contains a white background (marked in red outline in the image). This line must take on the color of the panel.

**(PR View)**

<img width="714" alt="Screenshot 2021-11-01 at 14 45 39" src="https://user-images.githubusercontent.com/15719418/139690337-7449027b-79ca-4325-8ae5-85ed1942ceee.png">

**(Final version)**

![image](https://user-images.githubusercontent.com/15719418/139690879-f1b80b8f-da8a-44d7-9e65-54a8363214bf.png)

3. The tab appearance of Active and Selected does not match the design. Will be linked to issue #4273, or is it a separate thing? As @aarmoa mentioned in today's call (November 1) (00:37:30).

**(UI Status tabs)**

![image](https://user-images.githubusercontent.com/15719418/139691675-93ee6b2b-8802-4644-a062-5d48afb888cc.png)

Nullably commented:
@luisfilipeoliveira 
> 1. The paddings in the log-pane are missing, but I believe that a dependency on this PR: [feat/add padding to panes #4579](https://github.com/CoinAlpha/hummingbot/pull/4579) is missing. Am I correct? As @aarmoa mentioned in today's call (November 1).

Just fixed. The padding should be there on the right pane as well.
> 2. The tab line contains a white background (marked in red outline in the image). This line must take on the color of the panel.

Are you using Windows? I just fixed this as well, but have not tested it on Windows yet.

> 1. The tab appearance of Active and Selected does not match the design. Will be linked to issue [Set a graphical appearance to command parameters #4273](https://github.com/CoinAlpha/hummingbot/issues/4273), or is it a separate thing? As @aarmoa mentioned in today's call (November 1) (00:37:30).

Fixed

luisfilipeoliveira commented:
@Nullably 

> Are you using Windows? I just fixed this as well, but have not tested it on Windows yet.

No, I'm using Macbook.

**Tip for keyboard shortcuts when navigating between tabs:**

- Alt (Or Command) + 1 - Switch to Tab1
- Alt (Or Command) + 2 - Switch to Tab2
- Alt (Or Command) + 3 - Switch to Tab3 
- ...
- And also Alt (Or Command) + left-arrowkey or Alt (Or Command) + rigth-arrowkey - to switch from left to right through your tabs and vice versa.

Tell me if these shortcuts are going to conflict with any hummingbot client requirement.

---

In terms of UI Design, everything seems perfect. I just wanted to see if you can give a space character to the right of the "x" (in the case of the tab that is enabled)

**PR Capture:**

<img width="436" alt="Screenshot 2021-11-02 at 16 04 05" src="https://user-images.githubusercontent.com/15719418/139894130-da019937-7ba4-40f2-9782-4a33177bc505.png">

**Ui Version**

![image](https://user-images.githubusercontent.com/15719418/139896848-60bdaa96-fd82-4ce2-856e-32d27ff39b6b.png)

Nullably commented:
> **Tip for keyboard shortcuts when navigating between tabs:**
> 
> * Alt (Or Command) + 1 - Switch to Tab1
> * Alt (Or Command) + 2 - Switch to Tab2
> * Alt (Or Command) + 3 - Switch to Tab3
> * ...
> * And also Alt (Or Command) + left-arrowkey or Alt (Or Command) + rigth-arrowkey - to switch from left to right through your tabs and vice versa.

There is a lot of restriction on what keys we can use. At the moment I have set it to Command (or Alt) + arrow keys to navigate left and right. If that doesn't work for you, you can try esc + arrow keys.
I don't think we should do the Alt + Number keys, as this might not work for many platforms and can create more support and maintenance issues.
 
> In terms of UI Design, everything seems perfect. I just wanted to see if you can give a space character to the right of the "x" (in the case of the tab that is enabled)

Done.

mhrvth commented on hummingbot/client/hummingbot_application.py:
```diff
@@ -269,3 +278,19 @@ def _initialize_notifiers(self):
+    def init_command_tabs(self) -> Dict[str, CommandTab]:
+        command_tabs: Dict[str, CommandTab] = {}
+        for mod_name, module in tab_package.__dict__.items():
+            for x in dir(module):
```
the name of the 'x' variable should probably be more specific

mhrvth commented on hummingbot/client/tab/__init__.py:
```diff
@@ -0,0 +1,9 @@
+import os
+import importlib
```
imports should be in alphabetical order as agreed two weeks ago

mhrvth commented on hummingbot/client/tab/data_types.py:
```diff
@@ -0,0 +1,18 @@
+import asyncio
+from dataclasses import dataclass
+from typing import Type, Optional
```
imports should be in alphabetical order as agreed two weeks ago

mhrvth commented on hummingbot/client/tab/dummy.pyx:
```diff
@@ -0,0 +1,2 @@
+cdef class dummy():
+    pass
```
What's the point of these classes? I can't find them being used anywhere.

mhrvth commented on hummingbot/client/tab/order_book_tab.py:
```diff
@@ -0,0 +1,71 @@
+import pandas as pd
+import asyncio
```
imports should be in alphabetical order as agreed two weeks ago

mhrvth commented on hummingbot/client/tab/tab_base.py:
```diff
@@ -0,0 +1,31 @@
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from hummingbot.client.hummingbot_application import HummingbotApplication
+from abc import (
```
imports should be in alphabetical order as agreed two weeks ago

mhrvth commented on hummingbot/client/ui/hummingbot_cli.py:
```diff
@@ -1,7 +1,8 @@
 
 import asyncio
-from typing import Callable, Optional
+import threading
```
imports should be in alphabetical order as agreed two weeks ago

mhrvth commented on hummingbot/client/ui/layout.py:
```diff
@@ -1,11 +1,14 @@
-from os.path import dirname, join, realpath
+from os.path import join, realpath, dirname
+from typing import Dict
+from prompt_toolkit.layout import Dimension
```
imports should be in alphabetical order as agreed two weeks ago

mhrvth commented on hummingbot/client/ui/style.py:
```diff
@@ -120,6 +120,8 @@ def hex_to_ansi(color_hex):
     "primary":                    "#1CD085",  # noqa: E241
     "warning":                    "#93C36D",  # noqa: E241
     "error":                      "#F5634A",  # noqa: E241
+    "tab_button.focused":         "bg:#1CD085 #000000",  # noqa: E241
```
I don't think this belongs here. hex_to_ansi() is just for converting individual color values from hex to windows ANSI codes because Windows command line is very limited. Styles of specific components are defined on the bottom of this file.

mhrvth replied:
... and the top functions apply the configurable settings to the styles. The reason for this is that there's no 1-to-1 correspondence between user defined settings and the actual styles of the components, and at the same time the user defined settings don't represent just simple colors, they're like styles applied to the actual styles, creating "connections" between e.g. foreground colors of multiple toolkit styles. You should define the base style of the component on the bottom and then in the top functions define how will the user settings affect it, and use the hex_to_ansi() function to convert the individual colors in hex chosen by the user for the particular OS.

mhrvth commented on hummingbot/client/ui/style.py:
```diff
@@ -136,4 +138,6 @@ def hex_to_ansi(color_hex):
     "primary":                    "#ansigreen",  # noqa: E241
     "warning":                    "#ansibrightyellow",  # noqa: E241
     "error":                      "#ansired",  # noqa: E241
+    "tab_button.focused":         "bg:#ansigreen #ansiblack",  # noqa: E241
```
I don't think this belongs here. hex_to_ansi() is just for converting individual color values from hex to windows ANSI codes because Windows command line is very limited. Styles of specific components are defined on the bottom of this file.

mhrvth commented on test/hummingbot/client/ui/test_style.py:
```diff
@@ -36,7 +36,10 @@ def test_load_style_unix(self, is_windows_mock):
                                  "warning": "#93C36D",
-                                 "error": "#F5634A"})
+                                 "error": "#F5634A",
+                                 "tab_button.focused": "bg:#1CD085 #000000",
```
This relates to the style.py comments

mhrvth commented on test/hummingbot/client/ui/test_style.py:
```diff
@@ -61,7 +64,10 @@ def test_load_style_windows(self, is_windows_mock):
                                  "warning": "#ansibrightyellow",
-                                 "error": "#ansired"})
+                                 "error": "#ansired",
+                                 "tab_button.focused": "bg:#ansigreen #ansiblack",
```
This relates to the style.py comments

mhrvth commented on test/hummingbot/client/ui/test_style.py:
```diff
@@ -83,7 +89,10 @@ def test_reset_style(self):
                                  "warning": "#93C36D",
-                                 "error": "#F5634A"})
+                                 "error": "#F5634A",
+                                 "tab_button.focused": "bg:#1CD085 #000000",
```
This relates to the style.py comments

luisfilipeoliveira commented:
@Nullably @aarmoa @mhrvth 

> There is a lot of restriction on what keys we can use. At the moment I have set it to Command (or Alt) + arrow keys to navigate left and right. If that doesn't work for you, you can try esc + arrow keys.

I don't appreciate the option of being the [esc] key so much. This in the hummingbot client is assumed to close the order book. In general, in the web community or application systems, it's to get out of the way we present ourselves.

> I don't think we should do the Alt + Number keys, as this might not work for many platforms and can create more support and maintenance issues.

I've been doing research and perceiving in terms of UX the best solution for us but paying attention to mac, windows, and Linux requirements. Tell me if what I'm going to present to you makes any sense:

**Shortcut op1:**
- Mac: [command] + [option] + [arrows keys]
- Windows: [control] + [alt] + [arrows keys]

**Shortcut op2:**
- Mac: [command] + [8] (left)
- Mac: [command] + [9] (right)
- Windows: [control] + [8]
- Windows: [control] + [9]

----

 A point I tested now.

You are not applying the `terminal-primary` color to the selected tab. When there is a color change update, in the case of the tab, there is no change at all.

![image](https://user-images.githubusercontent.com/15719418/140129800-2dd8d82c-b7c6-48d4-b31f-f1fd5001f6ee.png)

dennisocana commented:
Just wanted to drop a note here, ⌘ + arrow keys are for Spectacle app I'm using to organize my application windows. I'm thinking how that works with Hummingbot if we are to implement the first shortcut option.

https://user-images.githubusercontent.com/50150287/140239174-58ccc94c-6e41-460d-9fed-5a25ed027b02.mov

luisfilipeoliveira commented:
@dennisocana I just tested the Spectacle app you mentioned…so, you're right. Option 1 I recommended will create conflicts. 

Shortcut op2: would be the most functional for us. This is to avoid conflicting with shortcuts on windows and mac systems.

I will update one more option in my comment above.

Nullably commented:
@luisfilipeoliveira 
> A point I tested now.

> You are not applying the terminal-primary color to the selected tab. When there is a color change update, in the case of the tab, there is no change at all.

Can you give more specific requirement on this? 
Do you want the background colour of the tab to be the same color as the HB font color (terminal primary)? What about the foreground colour? And what colours are for the unselected tabs?

luisfilipeoliveira commented:
@Nullably What you've done in this new update seems fine to me.

- [ ] selected tab color (update color option according to settings — terminal primary) 
- [ ] keyboard shortcuts

Nullably commented on test/hummingbot/client/ui/test_style.py:
```diff
@@ -52,16 +55,19 @@ def test_load_style_windows(self, is_windows_mock):
-                                 "search": "#ansigray",
-                                 "search.current": "#ansigray",
-                                 "primary": "#ansigray",
+        style = Style.from_dict({"output-field": "bg:#ansiwhite #ansiwhite",
```
@mhrvth Since we added #ansiwhite to the hex convert function, I am not sure if this is an expected result from this test, can you please help check?

mhrvth replied:
Strange, before there was ansiwhite in the test, so it looks like someone removed it from the hex_to_ansi()

mhrvth replied:
after further examination I conclude that it was really missing in the function

mhrvth reviewed (approved):
LGTM

nikspz commented:
Upon cross-checking with Patrick **bot crashes after start on Win10/Mac Binary versions** throwing `FileNotFoundError` for `../hummingbot/client/tab`

Win10
![image](https://user-images.githubusercontent.com/83953535/140592437-d81727b4-367b-4fa3-ad32-461bc98b9225.png)
Mac
![image](https://user-images.githubusercontent.com/83953535/140592476-180bcc31-16fb-4427-b56f-d407dad0209f.png)
mov:
https://user-images.githubusercontent.com/83953535/140592528-2bc20bba-efe8-4237-a795-fd423a14f2ee.mov

rapcmia commented:
Tested on Docker build, switching to the next tab using `ctrl + N` works fine however when switching to the previous tab using `ctrl + P`, we have to press `P` twice to work

Nullably commented:
@nikspz @rapcmia 

> Upon cross-checking with Patrick bot crashes after start on Win10/Mac Binary versions throwing FileNotFoundError for ../hummingbot/client/tab

I have fixed this on the latest commit.

> Tested on Docker build, switching to the next tab using ctrl + N works fine however when switching to the previous tab using ctrl + P, we have to press P twice to work

https://forums.docker.com/t/ctrl-p-within-docker-container/5232
@aarmoa This is an outstanding known issue for dockers, I am not if we should implemented the suggested solution (the 3rd comment)

@mhrvth 
I have updated the tab/init.py file to list all tab classes there, there is no easy way to dynamically import all the sub modules of the tab package (note that py files are not present on binaries, so dynamic import on py files won't work). The other option is to install pkutils library, though I don't think we should go that far. What this means is a developer/user who creates a new tab will have to add their tab on the init file manually, and I think this is ok as long as we document it.

rxlxrxsx commented:
> Upon cross-checking with Patrick bot crashes after start on Win10/Mac Binary versions throwing FileNotFoundError for ../hummingbot/client/tab

This is confirmed fixed on the latest commit.

Just to add with the Ctrl + P issue on Docker: With the tab currently at order_book tab and I press Ctrl + P (+P because of the issue), instead of switching to tab_example it will switch directly to the log_pane tab, skipping other tabs that are open.

<img width="464" alt="Screen Shot 2021-11-08 at 3 41 44 PM" src="https://user-images.githubusercontent.com/64377055/140702493-0866c6b7-3d7e-47f2-ae50-dea831f9cc23.png">

mhrvth reviewed (dismissed):
LGTM

Nullably commented:
@CoinAlpha/qa-team @rxlxrxsx 
I have updated the tab left navigation to Ctl-b (for back) as suggested during stand-up.

> Just to add with the Ctrl + P issue on Docker: With the tab currently at order_book tab and I press Ctrl + P (+P because of the issue), instead of switching to tab_example it will switch directly to the log_pane tab, skipping other tabs that are open.

Can you please retest this.

@dennisocana @phbrgnomo 
Ctl-, ctl-. Ctl-[ are not registered.
See possible keys here (though not all of them works on every environment)
https://python-prompt-toolkit.readthedocs.io/en/master/pages/advanced_t
opics/key_bindings.html#binding-alt-something-option-something-or-meta-something

rxlxrxsx reviewed (dismissed):
Tested and confirmed working by QA.

dennisocana commented:
@Nullably ctrl + B and ctrl + N is good since they're just next to each other. Ctrl + B moves the tab to the left (previous) and Ctrl + N moves it to the right (next).

rxlxrxsx reviewed (approved):
Re-approving in behalf of QA.
